clients/py/sapphirepy/sapphire.py

```diff
@@ -218,60 +259,137 @@ def sapphire_middleware(
                         if pk:
                             manager.add(pk)
                     if not pk:
-                        raise RuntimeError('Could not retrieve callDataPublicKey!')
+                        raise RuntimeError("Could not retrieve callDataPublicKey!")
                     do_fetch = False
 
-                    c = _encrypt_tx_params(pk, params, method, w3, account)
-
-                    # We may encounter three errors here:
-                    #  'core: invalid call format: epoch too far in the past'
-                    #  'core: invalid call format: Tag verification failed'
-                    #  'core: invalid call format: epoch in the future'
-                    # We can only do something meaningful with the first!
-                    result = cast(RPCResponse, make_request(method, params))
-                    if result.get('error', None) is not None:
-                        error = result['error']
-                        if not isinstance(error, str) and error['code'] == -32000:
-                            if error['message'] == 'core: invalid call format: epoch too far in the past':
+                    if method == "eth_call" and params[0]["from"] and self.sapphire_account:
+                        # Get block number - use inline if to handle async vs sync
+                        block_number = (
+                            await self._w3.eth.block_number 
+                            if isinstance(self._w3, AsyncWeb3) 
+                            else self._w3.eth.block_number
+                        )
+                        
+                        # Get chain ID - inline if
+                        chain_id = (
+                            await self._w3.eth.chain_id
+                            if isinstance(self._w3, AsyncWeb3)
+                            else self._w3.eth.chain_id
+                        )
+                        
+                        # Get transaction count - inline if
+                        nonce = (
+                            await self._w3.eth.get_transaction_count(
+                                self._w3.to_checksum_address(params[0]["from"])
+                            )
+                            if isinstance(self._w3, AsyncWeb3)
+                            else self._w3.eth.get_transaction_count(
+                                self._w3.to_checksum_address(params[0]["from"])
+                            )
+                        )
+                        
+                        # Get block and extract hash - inline if
+                        block_hash = (
+                            (await self._w3.eth.get_block(block_number - 1))["hash"].hex()
+                            if isinstance(self._w3, AsyncWeb3)
+                            else self._w3.eth.get_block(block_number - 1)["hash"].hex()
+                        )
+                        
+                        signed_call_data = {
```

Comment:
Find appropriate type for `signed_call_data` (either inside `web3` package or define it yourself - should resemble to sth we use in https://github.com/oasisprotocol/sapphire-paratime/tree/main/clients/go).

Reply:
Added AttributeDict from web3:
`from web3.datastructures import (
    AttributeDict
)`